be/src/olap/rowset/segment_v2/page_io.cpp

```diff
@@ -0,0 +1,205 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "olap/rowset/segment_v2/page_io.h"
+
+#include <cstring>
+#include <string>
+
+#include "common/logging.h"
+#include "env/env.h"
+#include "gutil/strings/substitute.h"
+#include "olap/page_cache.h"
+#include "util/block_compression.h"
+#include "util/coding.h"
+#include "util/crc32c.h"
+#include "util/faststring.h"
+#include "util/runtime_profile.h"
+
+namespace doris {
+namespace segment_v2 {
+
+using strings::Substitute;
+
+Status PageIO::compress_page_body(const BlockCompressionCodec* codec,
+                                  double min_space_saving,
+                                  const std::vector<Slice>& body,
+                                  OwnedSlice* compressed_body) {
+    if (codec != nullptr) {
+        size_t uncompressed_size = Slice::compute_total_size(body);
+        size_t max_compressed_size = codec->max_compressed_len(uncompressed_size);
+        faststring buf;
+        buf.resize(max_compressed_size);
+        Slice compressed_slice(buf);
+        RETURN_IF_ERROR(codec->compress(body, &compressed_slice));
+        buf.resize(compressed_slice.get_size());
+
+        double space_saving = 1.0 - static_cast<double>(buf.size()) / uncompressed_size;
```

Comment:
Better to handle case when uncompressed_size is 0.

Reply:
OK